### ISSUE TITLE
JavaFile.writeTo(Path) should return output Path

### ISSUE THREAD

**Issue by koritakoa** [square/javapoet#691]
`JavaFile.writeTo(Path)` does some work to calculate the actual Path to which the source is written. There is no way to find what that output Path is without duplicating the logic. `writeTo(Path)` should return the actual output `Path` so we don't have to. `writeTo(File)` should return a `File`.

If you don't want to change the method signature, please consider adding a comment to these methods that the signature should change in 2.x and add a separate method to return the output path taking a Path parameter.

**Comment by JakeWharton** [comment]
Makes sense.

**Comment by galderz** [comment]
Indeed, changing the method signature would break binary compatibility since changing return type has the effect of removing the old method and adding the new one.

I'm not sure though about adding a separate method that returns the output path. First, I couldn't think of a good name for such method and second, it felt like you're trying to find a way to define an subset of an operation that is meant to stay as internal detail.

Instead, I'd like to suggest adding two methods that would keep the API relatively consistent with the downside of slightly bloating the method names:

```
/** Writes this to {@code directory} as UTF-8 using the standard directory structure.
   * Returns the {@link Path} instance to which source is actually written. */
public Path writeToPath(Path directory) throws IOException { ... }
```

```
/** Writes this to {@code directory} as UTF-8 using the standard directory structure.
 * Returns the {@link File} instance to which source is actually written. */
public File writeToFile(File directory) throws IOException { ... }
```

I've created a PR (#701) with this proposal (note: no tests added yet).
